Update: Slack: Adding more info to Slack Open Channel command

RobErskine opened the pull request:
## Description
Hi there! I updated the Slack Search Command to pull more data from Slack, like the user's Title (if it exists), their current status in Slack, when they will be back, their timezone location, as well as their current local time.

### Current Search Command UI
![Arc-_2024-07-05 at 17 07@2x](https://github.com/raycast/extensions/assets/1895120/18b375f2-8384-42f2-829d-57d6b33bd4ee)

### New Search Command UI
![Arc-_2024-07-05 at 17 12@2x](https://github.com/raycast/extensions/assets/1895120/c92f0f79-bfdb-4739-8c5e-8b5573c76eb1)

## Screencast

https://github.com/raycast/extensions/assets/1895120/77361592-390b-4574-8afa-ad8da58e9c06

### Bonus
Adds Emojify to Unread Messages as well to help make messages with emojis more readable:
![Code-unread-messages tsx — slack_2024-07-05 at 17 20@2x](https://github.com/raycast/extensions/assets/1895120/afe5a20a-0bbd-4753-84a5-c9564820e857)


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
![Warp- sionsextensionsslack_2024-07-05 at 16 56@2x](https://github.com/raycast/extensions/assets/1895120/41f31286-d393-4559-8ef6-0406d54f1582)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

momme-rtf commented:
Hi @RobErskine 
Thanks for your contribution! I like your change for displaying the emojis properly.
Regarding the search: I personally like to have the list more focused and less overloaded with infos. Especially when working within a small team I rather do not care about titles and locations of team members. Within a bigger company I see that it can be helpful. Maybe we could make it configurable somehow what we show. But I'd also like to hear the opinion of other contributers. What do you think?

RobErskine commented:
Hi @momme-rtf, thanks for your thoughts. I can see your point about keeping in small and uncluttered. Personally my org on Slack is over 700 people all over the world, so looking people up and understanding where they are and what time it is for them is crazy useful for me.

I do like your suggestion of adding in a Preference for this though to keep this simple for folks that like the way it is currently. I was already thinking in the shower that I should probably add in a preference for how the time should be displayed (am/pm or 24 hour). 

How about this:
- Add in a Preference for whether to show time in am/pm or 24 hour
- Add in a Preference on whether or not to show the Timezone / current time for that team member
- I think it's useful to show the Slack status emoji, the message, and when the status expires. Do you think that would be a good addition to keep in, without a Preference to toggle on/off?

jfkisafk commented on extensions/slack/package.json:
```diff
@@ -75,6 +75,7 @@
     "@raycast/api": "^1.76.0",
     "@raycast/utils": "^1.16.0",
     "@slack/web-api": "^7.0.4",
+    "@types/node-emoji": "^2.1.0",
```
Question (non-blocking): Shouldn't the types be added to devDependencies?

jfkisafk commented on extensions/slack/src/search.tsx:
```diff
@@ -19,12 +30,33 @@ function Search() {
             <List.Item
               key={userId}
               title={name}
+              subtitle={title}
               icon={icon ? { source: icon, mask: Image.Mask.Circle } : Icon.Person}
+              accessories={[
+                {
+                  icon: convertSlackEmojiToUnicode(statusEmoji ?? ""),
+                  text: statusText ?? "",
+                  tooltip: statusExpiration ?? 0,
+                },
+                { icon: Icon.Globe, text: timezone.split("/")[1].replace(/_/g, " ") },
+                { icon: Icon.Clock, text: getCoworkerTime(timezone) },
+              ]}
```
Suggestion: As you already indicated, a preference to show the metadata might keep things backwards-compatible. Beyond that future-proofing with granular options for specific metadata like AM/PM vs 24-hr time is a good-to-have but not really necessary imo (but I could be wrong about this). If folks create an issue requiring specific metadata prefs like that, we can handle it with iterative PRs.

Also, accessory has a date option which takes `Date` object and shows relative info. Should we use that here with more info in the `tooltip` of that accessory?

RobErskine replied:
I attempted `Date` accessory but found that it wasn't super useful to show the time as "3hr" as it didn't portray what the time actually means related to your current time. If you can think of another way to make that work I'll take another stab.

jfkisafk replied:
Yeah I thought about it and it is not applicable to what we are trying to achieve here, you can resolve this!

jfkisafk commented on extensions/slack/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # Slack Changelog
 
+## [Update to Search command] - 2024-07-05 
```
```suggestion
## [Update to Search command] - {PR_MERGE_DATE} 
```

jfkisafk commented on extensions/slack/package.json:
Suggestion (non-blocking): Please add yourself to the contributors if you want.

RobErskine replied:
Just added! Feels great to be in good company

jfkisafk reviewed (commented):
LGTM other than the broader conversation we are having around preferences.

RobErskine commented:
👋🏼 hey @jfkisafk @momme-rtf,

I went ahead and added two additional preferences to the extension for Time Format as well as Displaying Additional Metadata in the Search Command:
![Raycast-Settings_2024-07-07 at 14 28@2x](https://github.com/raycast/extensions/assets/1895120/d47f1a80-dafd-4366-ae96-e5d88d53b3ef)

I also added in HH:MM to the Slack Status Tooltip so if someone is grabbing lunch, or at a doctor's appt on the same day, the return time will be displayed. This screenshot also has the "Show Extra Metadata" option turned off (which is the default).
![Raycast-Settings_2024-07-07 at 14 29@2x](https://github.com/raycast/extensions/assets/1895120/48487e56-4187-419e-8be0-6f8ee54e4182)

Here is what it looks like with the extra metadata turned on, as well as the 24 hour time format.
![Arc-_2024-07-07 at 14 43@2x](https://github.com/raycast/extensions/assets/1895120/d2db1c72-34df-462e-8fe0-0487fe175a5d)

Let me know if you have any other feedback!

momme-rtf commented:
Awesome @RobErskine 
The preferences that you've added look good to me :)

jfkisafk commented on extensions/slack/src/shared/client/WebClient.ts:
```diff
@@ -2,6 +2,11 @@ import { getPreferenceValues } from "@raycast/api";
 import { OAuthService } from "@raycast/utils";
 import { WebClient } from "@slack/web-api";
 
+export type Preferences = {
+  timeFormat: "hour12" | "hour24";
+  displayExtraMetadata: boolean;
+};
```
Nitpick (non-blocking): this is no longer needed and you can just use `Preferences` created in _raycast-env.d.ts_: [src](https://developers.raycast.com/api-reference/preferences#getpreferencevalues)

You will need to add this to tsconfig.json though:
https://github.com/raycast/extensions/blob/adda091532b4805defd86396fc2b6ddb2eae05f3/extensions/todoist/tsconfig.json#L4

jfkisafk commented on extensions/slack/package.json:
```diff
@@ -69,13 +70,40 @@
+      "default": "hour12"
+    },
+    {
+      "title": "Search Command",
+      "label": "Display extra metadata in search results",
+      "description": "Show team member's titles, time zone, and current time in search results",
+      "name": "displayExtraMetadata",
```
Question (non-blocking): shouldn't this be added to command preferences rather than global preferences?

RobErskine replied:
Woah, didn't realize you could do that. Update for that coming soon 
![Raycast-Settings_2024-07-07 at 20 34@2x](https://github.com/raycast/extensions/assets/1895120/9834a872-9b30-48c7-a62f-381633f1c643)

jfkisafk commented on extensions/slack/src/search.tsx:
```diff
@@ -46,17 +71,9 @@ function Search() {
               key={userId}
               title={name}
-              subtitle={title}
+              subtitle={displayExtraMetadata === true ? title : ""}
```
```suggestion
              subtitle={displayExtraMetadata ? title : undefined}
```

jfkisafk reviewed (approved):
Minor comments but lgtm 🚀

jfkisafk commented on extensions/slack/src/search.tsx:
```diff
@@ -4,9 +4,8 @@ import { withSlackClient } from "./shared/withSlackClient";
-import { Preferences } from "./shared/client/WebClient";
 
-const { displayExtraMetadata } = getPreferenceValues<Preferences>();
+const { displayExtraMetadata } = getPreferenceValues();
```
```suggestion
const { displayExtraMetadata } = getPreferenceValues<Preferences.Search>();
```

jfkisafk commented on extensions/slack/src/shared/utils.ts:
```diff
@@ -3,10 +3,9 @@ import { showFailureToast } from "@raycast/utils";
 
 function getTimeLocale(): string {
-  const timeFormat = getPreferenceValues<Preferences>().timeFormat;
+  const timeFormat = getPreferenceValues().timeFormat;
```
```suggestion
  const timeFormat = getPreferenceValues<Preferences>().timeFormat;
```

jfkisafk reviewed (approved):
Minor suggested changes for type casting the preferences from _raycast-env.d.ts_, but not required. Ty for this awsum addition 🫶

jfkisafk commented:
Please rebase on top of https://github.com/raycast/extensions/pull/13039 and see if everything works as before? This will add a new command and the previous `Search` command name has changed.

RobErskine commented:
👋🏼 hi there.

This is pretty poor (or good?) timing on my end but I was working on an update to the ~Search~ Open Channel command just as @thomaslombart published [the latest Slack extension updates](https://twitter.com/thomas_lombart/status/1810313024345166160).

This PR adds Metadata to the Open Channel command to show Slack member's current status, timezone, as well as their current time. Also added in the option to display time in AM/PM or 24 hour time.

With the Metadata set to `false` (this is the default to keep functionality as close to what it is now):
![Arc-_2024-07-08 at 12 34@2x](https://github.com/raycast/extensions/assets/1895120/378d9878-6613-4174-9334-f8ced3202729)

With the Metadata turned on (with 24 hour time): 
![Arc-_2024-07-07 at 14 43@2x](https://github.com/raycast/extensions/assets/1895120/3fb26c05-c30a-4da4-a189-1b428ead9fd6)

thomaslombart commented:
Hey Rob, I'll review the PR next week 🙂

thomaslombart commented:
It seems other extension changes have slipped in while merging/rebasing. Could you fix it before I review your PR? 🙂

RobErskine commented:
@thomaslombart sorry about that; not sure how that happened while I was rebasing. I have merged `main` back in and all looks correct now. Mind giving it a look now?

jfkisafk commented on extensions/slack/src/shared/utils.ts:
```diff
@@ -1,8 +1,18 @@
+import * as emoji from "node-emoji";
+
+function getTimeLocale(): string {
+  const timeFormat = getPreferenceValues<Preferences>().timeFormat;
```
```suggestion
  const { timeFormat } = getPreferenceValues<Preferences>();
```

thomaslombart commented on extensions/slack/src/search.tsx:
```diff
@@ -1,11 +1,46 @@
+) {
+  const searchMetadata: Array<{ icon: string | Icon; text: string; tooltip?: string | null | undefined }> = [
+    {
+      icon: convertSlackEmojiToUnicode(statusEmoji ?? ""),
```
What about custom icons in Slack? I guess they won't be shown, right?

RobErskine replied:
Yeah unfortunately those still come through as a plain string to whatever is set in Slack. My team has one that I use all the time, `:thinking_rotate:`, which will just come in as `:thinking_rotate:`, like emojis used to. Maybe we can try to figure out how to get these to show up somehow? 
![:thinking_rotate:](https://emoji.slack-edge.com/T01JAPU340P/thinking_rotate/7a69dc59fdcf22fa.gif)

thomaslombart replied:
I guess we'll need to fetch custom emojis over [here](https://api.slack.com/methods/emoji.list), and then find the corresponding image. You can use `useCachedPromise` or `useFetch` as well for that. I guess there'll be a small delay for the user's statuses but I don't think that's a huge problem.

thomaslombart replied:
By the way, if you don't have time for that, we can already ship it as is an improve it later. That's more of a small improvement rather than a blocking issue.

thomaslombart commented on extensions/slack/src/shared/client/SlackClient.ts:
```diff
@@ -56,10 +62,31 @@ export class SlackClient {
+                minute: "2-digit",
+              });
+
+              statusExpirationDate = `Back on ${statusExpirationDate} at ${statusExpirationTime}`;
```
Would it be possible to format the date in a way that's more human-readable (such as [formatRelative](https://date-fns.org/v3.6.0/docs/formatRelative) from `date-fns`)

RobErskine replied:
This was a great suggestion. I think this is what Slack does as well. 
![date-fns - modern JavaScript date utility library-2024-07-18-21 52 37@2x](https://github.com/user-attachments/assets/6cda90b0-b8b0-4fb0-b5c3-1f0ef5a41fc8)

thomaslombart reviewed (commented):
Great improvements! 👏

I left a few comments.

RobErskine commented:
@thomaslombart thanks for the feedback. I've incorporated your feedback and I think it's better for it. Please take a look when you have the time

jfkisafk commented on extensions/slack/src/search.tsx:
```diff
@@ -1,11 +1,46 @@
+
+function getCoworkerTime(coworkerTimeZone: string): string {
+  const time = new Date();
+  return new Intl.DateTimeFormat(getTimeLocale(), {
```
Question (non-blocking): Can we not use `date-fns` for this formatting as well? Trying to see if we can remove the time preference altogether.

RobErskine replied:
That just might work, @jfkisafk. Removed the preference and replaced it with a +/- in time that shows hour and minutes if available: 
![search tsx — slack-2024-07-22-00 11 36@2x](https://github.com/user-attachments/assets/68767bc3-919f-413a-8772-c3452291c245)

Adds a little more of a cognitive load on the user but I think it's worth it

jfkisafk replied:
This looks good to me. Thanks for removing the additional preference.

jfkisafk commented on extensions/slack/package.json:
```diff
@@ -86,13 +98,15 @@
     "@raycast/utils": "^1.16.0",
     "@slack/web-api": "^7.0.4",
     "date-fns": "^3.6.0",
+    "date-fns-tz": "^3.1.3",
```
Question (non-blocking): Is this dependency required? I don't see any usage.

thomaslombart commented on extensions/slack/src/search.tsx:
```diff
@@ -1,11 +1,50 @@
+
+const { displayExtraMetadata } = getPreferenceValues<Preferences.Search>();
+
+function getCoworkerTime(coworkerTimeZone: string): string {
+  const localTime = new Date();
+  const coworkerTime = new Date(localTime.toLocaleString("en-US", { timeZone: coworkerTimeZone }));
+
+  const diffInMinutes = Math.round((coworkerTime.getTime() - localTime.getTime()) / (1000 * 60));
+  const diffHours = Math.floor(Math.abs(diffInMinutes) / 60);
+  const diffMinutes = Math.abs(diffInMinutes) % 60;
+
+  const formattedDiff = `${diffInMinutes >= 0 ? "+" : "-"}${diffHours}h${diffMinutes !== 0 ? ` ${diffMinutes}m` : ""}`;
+
+  return formattedDiff;
+}
```
You can probably use `date-fns` functions for that as well.

RobErskine commented:
I'm traveling for work but will be able to get to these last pieces of feedback tonight or tomorrow!

RobErskine commented:
@thomaslombart @jfkisafk Thanks for the feedback and apologies on the delay. I have updated the `getCoworkerTime` function to use `date-fns`and `date-fns-tz` to eliminate the need for an additional setting.

![search tsx — slack-2024-07-26-23 35 42@2x](https://github.com/user-attachments/assets/6ed135ea-7ab9-4f06-b297-81e9f10a8215)

![search tsx — slack-2024-07-26-23 36 05@2x](https://github.com/user-attachments/assets/0d8ac13f-0ef4-4dd4-895e-57c7ec69959f)

Now it only the time difference shows up if the there is an actual time difference. All time is also shown in hours as `-1h`, `+3h`, `+9.5h`, etc. This keeps the UI ubiquitous and simple. 

Let me know if you have any other questions or concerns!

thomaslombart reviewed (approved):
Awesome update! 🙌
